text/0000-dynamic-attributes.md

```diff
@@ -0,0 +1,91 @@
+---
+title: Passing Dynamic Attributes/Properties
+status: DRAFTED
+created_at: 2021-07-05
+updated_at: YYYY-MM-DD
+champion: Mohammed Abdul Sattar (abdulsattar)
+pr: https://github.com/salesforce/lwc-rfcs/pull/52
+---
+
+# Passing Dynamic Attributes/Properties
+
+## Summary
+This proposal adds mechanism to pass a dynamic attributes or properties bag to elements in
+an LWC template.
+
+## Example
+```js
+// x/myComponent.js
+export default class MyComponent extends LightningElement {
+    childProps = { 
+        name: 'My Child',
+        style: 'width: 50px',
+        class: 'my-class',
+        onclick: this.handleClick,
+    };
+    
+    handleClick() {
+        console.log('Clicked')
+    }
+}
+```
+```html
+// x/myComponent.html
+<template>
+    <x-child lwc:spread={childProps}></x-child>
+</template>
+```
+
+The new `lwc:spread` directive passes `childProps` as properties/attributes to `x-child`.
+
+## Motivation
+
+In LWC, there is no way to pass props/attributes to a child component when the component author
+doesn't know in advance all the props/attributes that are needed.
+Adding this feature can be helpful in creating wrapper components that extend the functionality
+of an existing component while being oblivious to all the props that component requires.
```

Comment:
I am not sold on how this proposal would make it easier to create wrapper components to extends existing component capabilities. This proposal solves the issue when you want to pass a set of properties/attributes instead of discrete bindings via the template. But it still requires the wrapper component to add all the public properties to be exposed via `@api`.

Reply:
Here is an example to better explain the point above. Let say that you want to extend the functionality of an `x-button` component with 3 public properties:

```js
// x-button.js

import { LigthningElement, api } from 'lwc';

export default class Buttton extends LightningElement {
  @api label;
  @api color; 
  @api active;
}
```

In the wrapper button component template you can use the newly introduced directive to pass down the public property it receives:

```html
<!-- x-wrapper-button.html -->
<template>
  <x-button lwc:spread={props}></x-button>
</template>
```

The main limitation with this proposal is that `x-wrapper-button` has to define the same set of public properties `x-button` defines for this work. This greatly limits the capability of extending capabilities for complex components like `lightning-input` with 50+ public properties and methods.

```js
// x-button.js

import { LigthningElement, api } from 'lwc';

export default class WrapperButtton extends LightningElement {
  // Indentical public properties than in x-button
  @api label;
  @api color; 
  @api active;
}
```

Reply:
Right, the only situation I see this helps is when we use `lwc:dynamic` (which is a totally different proposal).

Reply:
What if you already have map of the values in the wrapper component? For example you're building a `UL` using `for:each` to loop over an array of objects and construct `LI`s. If the `LI` contains a component expecting the values from the object, it would be so much nicer to spread the object into the internal component. Currently one must list each prop, which makes adding public properties to the inner component a bigger hassle than it needs to be.

```
<ul>
    <template for:each={items} for:item="item">
    <li>
        <some-component lwc:spread={item}></some-component>
    </li>
    </template>
</ul>
```

Reply:
@jonskeen what you mentioned is syntactic sugar. It's not like it's impossible to do it currently, it's just that `lwc:spread` will make passing those attributes a little easier. `lwc:spread` also comes at a cost: we can't do static analysis on `some-component` to see which properties are being passed to it.